pkgs/dart_model/lib/src/type_system.dart

```diff
@@ -61,6 +61,14 @@ final class StaticTypeSystem {
     return Scope.none.run(() => _isSubtype(a, b));
   }
 
+  /// Returns the super type of the type referred to by [name], which must be
+  /// part of the model.
+  ///
+  /// TODO(davidmorgan): is this the right place to provide access to supertypes?
+  QualifiedName supertypeOf(QualifiedName name) {
+    return _constructSuperTypes(_lookupNamed(name.asString), []).first.name;
```

Comment:
For types, I think this is the right place to put it. A `QualifiedName` alone is not really a type though, interface types would also need the instantiation. And this method answers class hierarchy queries, which are not exactly the same as type hierarchy queries:

1. There is no designated "primary" supertype in the type hierarchy since it's not relevant when checking for subtype relationships. Using `.first` here essentially relies on an implementation detail.
2. This generates an intermediate fully-resolved parent type when we only care about the name (so at the very least has to run in the none scope I think?)

I absolutely think that they're should be methods to resolve super types of interface _types_ in `TypeSystem`. I'm not sure if the `QualifiedName` is close enough to an interface type to also belong here though.
Especially if we care about the type of subclass relationship, maybe it's cleaner to add `extends`, `implements` and `mixesIn` properties on `Interface` as well?

In my mind, the kinds of question a type system should answer are "is this class, through whatever how many intermediates, a subtype of `JsonConverter`?". IMHO, "does this class extend from a class defining a `fromJson` constructor" is a different kind of question that is more related to the AST instead of a type model. But that might just be my opinion, so feel free to disagree or ignore this if it doesn't make sense :)

Reply:
All good points, thanks! I agree re: "what does this extend" being more like an AST question than a type model, I think moving this functionality to `Interface` is the right direction.

There's already a lot going on in this PR so added a TODO pointing to this comment + an item to the short term planning issue.